### PR TITLE
Mpileup speedup

### PR DESCRIPTION
**Pull request by jkbonfield**
This involves two main components.

1) Speeding up the probaln_glocal function to do less malloc/freeing, have better memory layout and to avoid so much cache pollution. (It still needs a lot of work there though.)

2) To add a callback mechanism to mpileup, permitting the caller to attach arbitrary additional information to pileup structs.  ABI breaking.

This is needed by my PR to the bcftools feature/mpileup branch.

### CODE REVIEW

**Comment by pd3** [comment]
+1000 for seeing this in htslib/develop soon. jkb's improvements to mpileup code depend on it

File: probaln.c [hunk 1]
```diff
@@ -110,23 +110,21 @@ int probaln_glocal(const uint8_t *_ref, int l_ref, const uint8_t *_query, int l_
...
             double e = (ref[k] > 3 || query[1] > 3)? 1. : ref[k] == query[1]? 1. - qual[1] : qual[1] * EM;
             set_u(u, bw, 1, k);
             fi[u+0] = e * bM; fi[u+1] = EI * bI;
             sum += fi[u] + fi[u+1];
         }
-        // rescale
         s[1] = sum;
-        set_u(_beg, bw, 1, beg); set_u(_end, bw, 1, end); _end += 2;
-        for (k = _beg; k <= _end; ++k) fi[k] /= sum;
+        // rescale
```
Comment (jmarshall) [inline probaln.c]: comment is now confusing

File: probaln.c [hunk 2]
```diff
@@ -232,14 +242,11 @@ int probaln_glocal(const uint8_t *_ref, int l_ref, const uint8_t *_query, int l_
         if (state) state[i-1] = max_k;
         if (q) k = (int)(-4.343 * log(1. - max) + .499), q[i-1] = k > 100? 99 : k;
 #ifdef _MAIN
-        fprintf(stderr, "(%.10lg,%.10lg) (%d,%d:%c,%c:%d) %lg\n", pb, sum, i-1, max_k>>2,
+        fprintf(stderr, "(%.10lg,%.10lg) (%d,%d:%c,%c:%d) %lg\n", 0.0/*pb*/, sum, i-1, max_k>>2,
```
Comment (jmarshall) [inline probaln.c]: Will have already been fixed by 8d957c074297e62400a08f05c3afee6df1bd4d27.

File: probaln.c [hunk 3]
```diff
@@ -85,16 +85,14 @@ int probaln_glocal(const uint8_t *_ref, int l_ref, const uint8_t *_query, int l_
...
-        if (is_backward) b[i] = calloc(bw2 * 3 + 6, sizeof(double));
-    }
-    s = calloc(l_query+2, sizeof(double)); // s[] is the scaling factor to avoid underflow
+    // Ideally these callocs would be mallocs + initialisation of the few bits needed.
+    f = calloc((l_query+1)*i_dim, sizeof(double));
+    if (is_backward) b = calloc((l_query+1)*i_dim, sizeof(double));
+    s = malloc((l_query+2)* sizeof(double)); // s[] is the scaling factor to avoid underflow
     // initialize qual
-    _qual = calloc(l_query, sizeof(float));
+    _qual = malloc(l_query* sizeof(float));
```
Comment (jmarshall) [inline probaln.c]: Spaces on both or neither side of `*` please :smile:

**Comment by jmarshall** [comment]
Apologies for the delay in reviewing this.

1. There's two separate components here: the _probaln.c_ changes, and the mpileup callbacks.  It would be good to have them as two separate pull requests; in fact, it would be reasonable to commit the _probaln.c_ changes on your own authority without much review.  (But happy to review for obvious problems.)

2. Seems that the second callback change reverts and reworks part of the first.  Please squash so we don't have to review the API false start.  Maybe please also have the `0` => `NULL` cosmetic change as a separate commit.

3. This prints out `DES1`/`DES2`/`DES3` debugging messages.  Seems this isn't actually ready for merging to **develop**.  Please resubmit when it's actually in a state for a full review and merge.

**Comment by jkbonfield** [comment]
Think I've dealt with all the comments.  Sorry about leaving the debugging in there - I don't know how that happened and I thought it was all ready, but I was in a rush to get it pushed out as I was on holiday a day later and I know it had other people waiting on it.

**Pull request merged**